base-containers/base/inginious_container_api/run_student.py

```diff
@@ -69,90 +68,20 @@ def run_student(cmd, container=None,
-        message = msgpack.loads(zmq_socket.recv(), use_list=False, strict_map_key=False)
-        assert message["type"] == "run_student_started"
-        student_container_id = message["container_id"]
-
-        # The socket only works in a ping-pong style so we need to send a dummy message
-        zmq_socket.send(msgpack.dumps({"type": "run_student_ask_retval", "socket_id": socket_id}, use_bin_type=True))
-
-        # Serve one and only one connection
-        connection, addr = server.accept()
-
-        # _run_student_intern should say hello
-        datagram = connection.recv(1)
-        assert datagram == b'H'
-
-        # send the fds and the command/workdir
-        connection.sendmsg([b'S'], [(socket.SOL_SOCKET, socket.SCM_RIGHTS, array.array("i", [stdin, stdout, stderr]))])
-        connection.send(msgpack.dumps({"command": cmd, "teardown_script": teardown_script, "working_dir": working_dir, "ssh": ssh, "user": user}))
-
-        # Allow to send signals
-        if signal_handler_callback is not None:
-            def receive_signal(signum_s):
-                signum_data = str(signum_s).zfill(3).encode("utf8")
-                connection.send(signum_data)
-            signal_handler_callback(receive_signal)
-
-        if ssh:  # The student_container will send id and password for ssh connection, transfer it to the agent
-            s = connection.recv(4)  # First 4 bytes are for the size
-            message_length = struct.unpack('!I', bytes(s))[0]
-            ssh_id = msgpack.loads(connection.recv(message_length))
-            if ssh_id["type"] == "ssh_student":
-                msg = {"type": "ssh_student", "ssh_user": ssh_id["ssh_user"], "ssh_key": ssh_id["password"], "container_id": student_container_id}
-                send_socket = zmq.asyncio.Context().socket(zmq.REQ)
-                send_socket.connect("ipc:///sockets/main.sock")
-                loop = asyncio.get_event_loop()
-                task = loop.create_task(_send_intern_message(send_socket, msg))
-                loop.run_until_complete(task)
-                loop.close()
-
-        # Wait for everything to end
-        # message = message from agent telling the student_container finished
-        message = msgpack.loads(zmq_socket.recv(), use_list=False, strict_map_key=False)
-
-        # Unlink unneeded files
-        try:
-            os.unlink(socket_path)
-            os.unlink(path)
-        except:
-            pass
-
+        server, socket_id, socket_path, path = create_student_socket(only_dockers)
+        zmq_socket, student_container_id = start_student_container(container, time_limit, hard_time_limit, memory_limit, share_network, socket_id, ssh, run_as_root)
+        connection = send_initial_command(socket_id, server, stdin, stdout, stderr, zmq_socket, student_container_id, cmd, teardown_script, working_dir, ssh, user, only_dockers)
+        allow_to_send_signals(signal_handler_callback, connection, student_container_id, only_dockers)
```

Comment:
Have you checked that it works when the command immediatly returns a signal?
Are you sure there are no race conditions here?

Reply:
Directly after sending the initial command (without waiting any answer from the student_container), we setup the handler to transfer signals to the student_container. This is exactly how it was already implemented previously (see https://github.com/UCL-INGI/INGInious/blob/3c8b5b333b817cb9741455dd7f0bf2cfaed2cfed/base-containers/base/inginious_container_api/run_student.py#L121).
I did not encounter any bugs up to now with that but if you have an interesting example in mind, feel free to send it to me :)